specs/accounting_reform/beacon-chain.md

```diff
@@ -0,0 +1,278 @@
+    if all(bits[1:3]) and old_previous_justified_checkpoint.epoch + 2 == current_epoch:
+        state.finalized_checkpoint = old_previous_justified_checkpoint
+    # The 1st/2nd/3rd most recent epochs are justified, the 1st using the 3rd as source
+    if all(bits[0:3]) and old_current_justified_checkpoint.epoch + 2 == current_epoch:
+        state.finalized_checkpoint = old_current_justified_checkpoint
+    # The 1st/2nd most recent epochs are justified, the 1st using the 2nd as source
+    if all(bits[0:2]) and old_current_justified_checkpoint.epoch + 1 == current_epoch:
+        state.finalized_checkpoint = old_current_justified_checkpoint
+```
+
+##### `get_flag_deltas`
+
+```python
+def get_flag_deltas(state: BeaconState, flag: uint8, numerator: uint64) -> Tuple[Sequence[Gwei], Sequence[Gwei]]:
+    rewards = [Gwei(0)] * len(state.validators)
+    penalties = [Gwei(0)] * len(state.validators)
+    unslashed_participating_indices = get_unslashed_participating_indices(state, flag, get_previous_epoch(state))
+
+    increment = EFFECTIVE_BALANCE_INCREMENT  # Factored out from balances to avoid uint64 overflow
+    unslashed_participating_increments = get_total_balance(state, unslashed_participating_indices) // increment
+    active_increments = get_total_active_balance(state) // increment
+    for index in get_eligible_validator_indices(state):
+        base_reward = get_base_reward(state, index)
+        if index in unslashed_participating_indices:
+            if is_in_inactivity_leak(state):
+                # Optimal participatition is fully rewarded to cancel the inactivity penalty
+                rewards[index] = base_reward * numerator // REWARD_DENOMINATOR
+            else:
+                rewards[index] = base_reward * numerator * unslashed_participating_increments
+                // (active_increments * REWARD_DENOMINATOR)
+        else:
+            penalties[index] = base_reward * numerator // REWARD_DENOMINATOR
+    return rewards, penalties
+```
+
+##### `get_inactivity_penalty_deltas`
+
+*Note*: The function `get_inactivity_penalty_deltas` is modified in the selection of matching target indices and the removal of `BASE_REWARDS_PER_EPOCH`.
+
+```python
+def get_inactivity_penalty_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence[Gwei]]:
+    """
+    Return inactivity reward/penalty deltas for each validator.
+    """
+    penalties = [Gwei(0) for _ in range(len(state.validators))]
+    if is_in_inactivity_leak(state):
+        reward_numerator_sum = sum(numerator for (_, numerator) in FLAGS_AND_NUMERATORS)
+        matching_target_indices = get_unslashed_participating_indices(state, TIMELY_TARGET_FLAG, get_previous_epoch(state))
+        for index in get_eligible_validator_indices(state):
+            # If validator is performing optimally this cancels all attestation rewards for a neutral balance
+            penalties[index] += get_base_reward(state, index) * reward_numerator_sum // REWARD_DENOMINATOR
```

Comment:
Given that you have to attest within one slot to get the head reward, and five epochs for the source reward, this seems even harsher than the previous computation (which was already heavily criticised for penalizing even optimally performing validators). In fact, getting the head right within one epoch is impossible at least 33% of the time, meaning that you are guaranteed to be down ~-1/6 already.

I'd suggest scaling this by 1/2, meaning you can be slightly positive still (which I think we have determined in the past is not a problem).

Reply:
You mean like in the suggestion I just made directly above?

Reply:
I even suggested being more lenient and dividing the whole `reward_numerator_sum` by 2.